### PR TITLE
Implement progress notifications

### PR DESCRIPTION
**Pull request by rubenporras**

### CODE REVIEW

**Comment by rubenporras** [comment]
@mickaelistria , this is still WIP. Do you think it goes in the right direction?

File: org.eclipse.lsp4e/src/org/eclipse/lsp4e/LanguageClientImpl.java [hunk 1]
```diff
@@ -83,6 +95,54 @@ public final void logMessage(MessageParams message) {
...
+	@Override
+	public CompletableFuture<Void> createProgress(WorkDoneProgressCreateParams params) {
+		String jobName = params.getToken().getLeft();
+		LinkedBlockingDeque<ProgressParams> blockingDeque = new LinkedBlockingDeque<ProgressParams>();
+		progressQueueMap.put(jobName, blockingDeque);
+
+		Job job = Job.create(jobName, new ICoreRunnable() {
+			@Override
+			public void run(IProgressMonitor monitor) throws CoreException {
+				while (true) {
```
Comment (mickaelistria) [inline org.eclipse.lsp4e/src/org/eclipse/lsp4e/LanguageClientImpl.java:109]: I'm a bit uncomfortable with `while (true)` what about `while((nextNotification = blockingQueue.take()) != null)` ?

**Review by mickaelistria — COMMENTED** [review]
Yes, I think it's the right direction.

File: org.eclipse.lsp4e/src/org/eclipse/lsp4e/LanguageClientImpl.java (hunk 1 above)
Comment (rubenporras) [reply org.eclipse.lsp4e/src/org/eclipse/lsp4e/LanguageClientImpl.java]: Thanks, that is much better.
I will continue working on it.

File: org.eclipse.lsp4e/src/org/eclipse/lsp4e/LanguageClientImpl.java [hunk 2]
```diff
@@ -83,6 +94,51 @@ public final void logMessage(MessageParams message) {
...
+				try {
+					while ((nextNotification = blockingQueue.take()) != null) {
+						WorkDoneProgressNotification nextProgressNotification = nextNotification.getValue().getLeft();
+						if (nextProgressNotification != null) {
+							WorkDoneProgressKind kind = nextProgressNotification.getKind();
+							if (kind == WorkDoneProgressKind.begin) {
+								monitor.beginTask(jobName, 100);
+							} else if (kind == WorkDoneProgressKind.report) {
+								WorkDoneProgressReport report = (WorkDoneProgressReport) nextProgressNotification;
+								monitor.worked(report.getPercentage());
```
Comment (bastiandoetsch) [inline org.eclipse.lsp4e/src/org/eclipse/lsp4e/LanguageClientImpl.java:118]: This doesn't take into account that in LSP, a percentage is reported, a progress monitor is accumulating progress. So the reported percentage should ideally take the already reported progress (how many of 100) into account and only add the delta to the progress monitor.

File: org.eclipse.lsp4e/src/org/eclipse/lsp4e/LanguageClientImpl.java [hunk 3]
```diff
@@ -83,6 +94,51 @@ public final void logMessage(MessageParams message) {
...
+						}
+					}
+				} catch (InterruptedException e) {
+					LanguageServerPlugin.logError(e);
+					Thread.currentThread().interrupt();
+				}
+			}
+		});
+		job.schedule();
+		return CompletableFuture.completedFuture(null);
```
Comment (bastiandoetsch) [inline org.eclipse.lsp4e/src/org/eclipse/lsp4e/LanguageClientImpl.java:132]: Cancellations would be a nice addition :)

File: org.eclipse.lsp4e/src/org/eclipse/lsp4e/LanguageClientImpl.java [hunk 4]
```diff
@@ -83,6 +94,51 @@ public final void logMessage(MessageParams message) {
...
+				ProgressParams nextNotification;
+				try {
+					while ((nextNotification = blockingQueue.take()) != null) {
+						WorkDoneProgressNotification nextProgressNotification = nextNotification.getValue().getLeft();
+						if (nextProgressNotification != null) {
+							WorkDoneProgressKind kind = nextProgressNotification.getKind();
+							if (kind == WorkDoneProgressKind.begin) {
+								monitor.beginTask(jobName, 100);
+							} else if (kind == WorkDoneProgressKind.report) {
+								WorkDoneProgressReport report = (WorkDoneProgressReport) nextProgressNotification;
```
Comment (bastiandoetsch) [inline org.eclipse.lsp4e/src/org/eclipse/lsp4e/LanguageClientImpl.java:117]: Would also be nice to have the message field analysed to create subtasks.

File: org.eclipse.lsp4e/src/org/eclipse/lsp4e/LanguageClientImpl.java [hunk 5]
```diff
@@ -83,6 +94,51 @@ public final void logMessage(MessageParams message) {
 		CompletableFuture.runAsync(() -> ServerMessageHandler.logMessage(wrapper, message));
 	}
 
+	private Map<String, BlockingQueue<ProgressParams>> progressQueueMap = new HashMap<>();
+
+	@Override
+	public CompletableFuture<Void> createProgress(WorkDoneProgressCreateParams params) {
+		String jobName = params.getToken().getLeft();
```
Comment (bastiandoetsch) [inline org.eclipse.lsp4e/src/org/eclipse/lsp4e/LanguageClientImpl.java:101]: Using the `token` as `jobName` may cause some jobs to be having a weird UX (e.g. our Language Server uses UUIDs as progress tokens). Maybe put at least the Language Server name in front of the token or use the message field in `WorkDoneProgressCreateParams`?

**Review by bastiandoetsch — COMMENTED** [review]
Looks like it's going to be well usable for us :), good work! Please consider adding test cases for the different scenarios, I didn't see any in the PR.

**Comment by rubenporras** [comment]
@bastiandoetsch , thanks for the review. Now it is me who is busy preparing for an internal release. Most probably I will not have time to work on this in the next days.

**Comment by rubenporras** [comment]
Hi @bastiandoetsch, I will start working on this again next week. Cheers

File: org.eclipse.lsp4e/src/org/eclipse/lsp4e/LanguageClientImpl.java (hunk 3 above)
Comment (rubenporras) [reply org.eclipse.lsp4e/src/org/eclipse/lsp4e/LanguageClientImpl.java]: done

File: org.eclipse.lsp4e/src/org/eclipse/lsp4e/LanguageClientImpl.java [hunk 6]
```diff
@@ -83,6 +94,51 @@ public final void logMessage(MessageParams message) {
...
+		});
+		job.schedule();
+		return CompletableFuture.completedFuture(null);
+	}
+
+	@Override
+	public void notifyProgress(ProgressParams params) {
+		String jobName = params.getToken().getLeft();
+		BlockingQueue<ProgressParams> progress = progressQueueMap.get(jobName);
+		progress.add(params);
```
[the remark this answers is not in the document]
Comment (rubenporras) [reply org.eclipse.lsp4e/src/org/eclipse/lsp4e/LanguageClientImpl.java]: done

File: org.eclipse.lsp4e/src/org/eclipse/lsp4e/LanguageClientImpl.java (hunk 5 above)
Comment (rubenporras) [reply org.eclipse.lsp4e/src/org/eclipse/lsp4e/LanguageClientImpl.java]: done

File: org.eclipse.lsp4e/src/org/eclipse/lsp4e/LanguageClientImpl.java (hunk 2 above)
Comment (rubenporras) [reply org.eclipse.lsp4e/src/org/eclipse/lsp4e/LanguageClientImpl.java]: Thanks for the hint.

**Comment by rubenporras** [comment]
@bastiandoetsch, it has been of great help to have your work to look at. Thanks a lot!

File: org.eclipse.lsp4e/src/org/eclipse/lsp4e/ProgressManager.java [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+import org.eclipse.lsp4j.WorkDoneProgressBegin;
+import org.eclipse.lsp4j.WorkDoneProgressCancelParams;
+import org.eclipse.lsp4j.WorkDoneProgressCreateParams;
+import org.eclipse.lsp4j.WorkDoneProgressEnd;
+import org.eclipse.lsp4j.WorkDoneProgressKind;
+import org.eclipse.lsp4j.WorkDoneProgressNotification;
+import org.eclipse.lsp4j.WorkDoneProgressReport;
+import org.eclipse.lsp4j.services.LanguageServer;
+
+public class ProgressManager {
```
[the remark this answers is not in the document]
Comment (rubenporras) [reply org.eclipse.lsp4e/src/org/eclipse/lsp4e/ProgressManager.java]: done.

File: org.eclipse.lsp4e/src/org/eclipse/lsp4e/LSPProgressManager.java [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *  Rubén Porras Campo (Avaloq Evolution AG) - initial implementation
+ *******************************************************************************/
+package org.eclipse.lsp4e;
```
Comment (mickaelistria) [inline org.eclipse.lsp4e/src/org/eclipse/lsp4e/LSPProgressManager.java:12]: Would it be possible to have this in a `org.eclipse.lsp4e.progress` package?

Comment (rubenporras) [reply org.eclipse.lsp4e/src/org/eclipse/lsp4e/LSPProgressManager.java]: done

**Comment by rubenporras** [comment]
Hi @mickaelistria , anything else or shall I merge the PR

**Comment by rubenporras** [comment]
Thanks

**Comment by jcompagner** [comment]
this case seem to have caused some hanging jobs: https://github.com/eclipse/lsp4e/issues/213

i guess an "end" event never comes?

**Pull request merged**